xmtp_db/src/encrypted_store/conversation_list.rs

```diff
@@ -114,53 +114,48 @@ impl DbConnection {
-
-                self.raw_query_read(|conn| query.load::<ConversationListItem>(conn))?
-            } else {
-                // Only include the specified states
-                let query = query
-                    .inner_join(
-                        consent_dsl::consent_records.on(sql::<diesel::sql_types::Text>(
-                            "lower(hex(conversation_list.id))",
-                        )
-                        .eq(consent_dsl::entity)),
-                    )
-                    .filter(consent_dsl::state.eq_any(consent_states.clone()))
-                    .select(conversation_list::all_columns())
-                    .order(conversation_list_dsl::created_at_ns.asc());
-
-                self.raw_query_read(|conn| query.load::<ConversationListItem>(conn))?
-            }
-        } else {
-            // Handle the case where `consent_states` is `None`
+        let effective_consent_states = match &consent_states {
+            Some(states) => states.clone(),
+            None => vec![ConsentState::Allowed, ConsentState::Unknown],
+        };
+
+        let includes_unknown = effective_consent_states.contains(&ConsentState::Unknown);
+        let includes_all = effective_consent_states.len() == 3;
+
+        let filtered_states: Vec<_> = effective_consent_states
+            .iter()
+            .filter(|state| **state != ConsentState::Unknown)
+            .cloned()
+            .collect();
+
+        let mut conversations = if includes_all {
+            // No filtering at all
             self.raw_query_read(|conn| query.load::<ConversationListItem>(conn))?
+        } else if includes_unknown {
+            // LEFT JOIN: include Unknown + NULL + filtered states
+            let left_joined_query = query
+                .left_join(consent_dsl::consent_records.on(
+                    sql::<diesel::sql_types::Text>("lower(hex(conversation_list.id))").eq(consent_dsl::entity),
+                ))
+                .filter(
+                    consent_dsl::state
+                        .is_null()
+                        .or(consent_dsl::state.eq(ConsentState::Unknown))
+                        .or(consent_dsl::state.eq_any(filtered_states.clone())),
+                )
+                .select(conversation_list::all_columns());
+
+            self.raw_query_read(|conn| left_joined_query.load::<ConversationListItem>(conn))?
+        } else {
+            // INNER JOIN: strict match only to specific states (no Unknown or NULL)
+            let inner_joined_query = query
+                .inner_join(consent_dsl::consent_records.on(
+                    sql::<diesel::sql_types::Text>("lower(hex(conversation_list.id))").eq(consent_dsl::entity),
+                ))
+                .filter(consent_dsl::state.eq_any(filtered_states.clone()))
+                .select(conversation_list::all_columns());
+
+            self.raw_query_read(|conn| inner_joined_query.load::<ConversationListItem>(conn))?
```

Comment:
I'm really at a loss here for how to preserve order while filtering for consent. I feel like I've tried a million different queries @insipx @mchenani do either of you have ideas? In my head this should just work? I don't know why the order is getting moved and the test fails but if I run it 5 times in a row it will pass so it's for sure ordering randomly.

Reply:
Do we need a `.order(conversation_list_dsl::created_at_ns.asc())`?

Reply:
I added that in and the tests still failed. Because this function in particular from my understanding orders the conversations my message last sent at ns?

Reply:
This was the fix!
`.order(sql::<BigInt>("COALESCE(sent_at_ns, created_at_ns) DESC"))`